File: go/data-access.go
package main

import (
	"sync"
)

var mutex = &sync.Mutex{}
var commandDependenciesMap = make(map[string][]string)

// IsIndexed returns true if the package name is in the map, and false if it's not
func IsIndexed(packageName string) bool {
	mutex.Lock()
	result := commandDependenciesMap[packageName] != nil
	mutex.Unlock()
	return result
}

// IsDependedOn checks whether a package name is current dependend on and returns a corresponding boolean
func IsDependedOn(packageName string) bool {
	mutex.Lock()
	for _, dependencies := range commandDependenciesMap {
		for i := range dependencies {
			if dependencies[i] == packageName {
				mutex.Unlock()
				return true
			}
		}
	}
	mutex.Unlock()
	return false
}

// IndexPackage indexes a package and it's dependencies in the data store
func IndexPackage(packageName string, dependencies []string) {
	mutex.Lock()
	commandDependenciesMap[packageName] = dependencies
	mutex.Unlock()
}

// RemovePackage removes a package and it's dependencies from the data store
func RemovePackage(packageName string) {
	mutex.Lock()
	delete(commandDependenciesMap, packageName)
	mutex.Unlock()
}

// WipeDataStore is a method exposed purely for testing
func WipeDataStore() {
	mutex.Lock()
	for packageName := range commandDependenciesMap {
		delete(commandDependenciesMap, packageName)
	}
	mutex.Unlock()
}

// GetDataStore is a method exposed purely for testing
func GetDataStore() map[string][]string {
	return commandDependenciesMap
}

File: node/src/data-access.ts

import { debug } from './logging';

const commandDependenciesMap = new Map<string, string[]>();

export async function isIndexed(packageName: string): Promise<boolean> {
  debug(`Starting isIndexed for package ${packageName}`)
  const result = commandDependenciesMap.has(packageName);
  debug(`Done with isIndexed for package ${packageName}`)
  return result;
}

export async function isDependedOn(packageName: string): Promise<boolean> {
  debug(`Starting isDependedOn for package ${packageName}`)
  let found = false;
  commandDependenciesMap.forEach((dependencies: string[], key: string) => {
    for (const dependency of dependencies) {
      if (dependency === packageName) {
        debug(`${packageName} isDependedOn by ${key}}`);
        found = true;
      }
    }
  });
  debug(`Done with isDependedOn for package ${packageName}`)
  return found;
}

export async function index(packageName: string, dependencies: string[]): Promise<void> {
  debug(`Starting index for ${packageName}`)
  commandDependenciesMap.set(packageName, dependencies);
  debug(`Done with index for ${packageName}`)
}

export async function remove(packageName: string): Promise<boolean> {
  debug(`Starting remove for ${packageName}`)
  const result = commandDependenciesMap.delete(packageName);
  debug(`Done with remove for ${packageName}`)
  return result;
}

// this method is used purely for testing
export function wipeDataStore() {
  const keysToDelete: string[] = [];
  commandDependenciesMap.forEach((value: string[], key: string) => {
    keysToDelete.push(key);
  });
  keysToDelete.forEach(key => commandDependenciesMap.delete(key));
}